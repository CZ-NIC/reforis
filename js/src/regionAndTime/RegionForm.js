/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import propTypes from 'prop-types';

import Select from '../common/bootstrap/Select';
import {TIMEZONES, COUNTRIES} from '../utils/timezones';

RegionForm.propTypes = {
    formData: propTypes.shape({
        region: propTypes.string,
        country: propTypes.string,
        city: propTypes.string,
    }),
    setFormValue: propTypes.func.isRequired,
};


RegionForm.defaultProps = {
    formData: {},
    formErrors: {},
    setFormValue: ()=>{},
};

export default function RegionForm({formData, formErrors, setFormValue, ...props}) {
    function getRegionChoices() {
        return Object.keys(TIMEZONES)
            .sort()
            .reduce((obj, key) => {
            obj[key] = _(key);
            return obj;
        }, {})
    }

    function getCountryChoices() {
        return Object.keys(TIMEZONES[formData.region])
            .sort()
            .reduce((obj, country_code) => {
            obj[country_code] = _(COUNTRIES[country_code]);
            return obj;
        }, {})
    }

    function getCityChoices() {
        const countries = TIMEZONES[formData.region];
        const cities = countries[formData.country] || countries[Object.keys(countries)[0]];
        return Object.keys(cities)
            .sort()
            .reduce((obj, city) => {
            obj[city] = _(city);
            return obj;
        }, {})
    }

    return <>
        <h4>{_('Region settings')}</h4>
        <p>{_('Please select the timezone the router is being operated in. Correct setting is required to display ' +
            'the right time and for related functions.')}</p>
        <Select
            choices={getRegionChoices()}
            label={_('Continent or ocean')}
            value={formData.region}

            onChange={setFormValue(
                value => {
                    const country = Object.keys(TIMEZONES[value])[0];
                    const city = Object.keys(TIMEZONES[value][country])[0];
                    return {
                        region: {$set: value},
                        country: {$set: country},
                        city: {$set: city},
                    }
                }
            )}
            {...props}
        />
        <Select
            choices={getCountryChoices()}
            label={_('Country')}
            value={formData.country || ''}

            onChange={setFormValue(
                value => ({
                    country: {$set: value},
                    city: {$set: Object.keys(TIMEZONES[formData.region][value])[0]},
                })
            )}

            {...props}
        />
        <Select
            choices={getCityChoices()}
            label={_('Timezone')}
            value={formData.city || ''}

            onChange={setFormValue(
                value => ({city: {$set: value}})
            )}

            {...props}
        />
    </>
}
