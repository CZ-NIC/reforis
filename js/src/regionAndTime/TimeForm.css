#ntpServersList {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
}

#ntpServersList > p {
    margin: 0.5rem;
}
