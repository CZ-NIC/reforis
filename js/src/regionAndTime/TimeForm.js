/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, {useEffect, useState} from 'react';
import PropTypes from 'prop-types';
import moment from 'moment/moment';

import Select from 'common/bootstrap/Select';
import DataTimeInput from 'common/bootstrap/DataTimeInput';
import {SpinnerElement} from 'common/bootstrap/Spinner';
import Button from 'common/bootstrap/Button';

import useNTPDate from './hooks';

// Foris backend ignore value after "."...
const TIME_FORMAT = 'YYYY-MM-DDTHH:mm:ss.0';

const TIME_SETTING_TYPE_CHOICES = {
    ntp: _('Via NTP'),
    manual: _('Manually'),
};

TimeForm.propTypes = {
    ws: PropTypes.object.isRequired,

    formData: PropTypes.shape({
        time_settings: PropTypes.shape({
            how_to_set_time: PropTypes.oneOf(['ntp', 'manual']),
            ntp_servers: PropTypes.arrayOf(PropTypes.string),
            time: PropTypes.oneOfType([
                PropTypes.objectOf(moment),
                PropTypes.string
            ]),
        }).isRequired,
    }).isRequired,

    formErrors: PropTypes.shape({
        time_settings: PropTypes.shape({
            time: PropTypes.string,
        }),
    }).isRequired,

    setFormValue: PropTypes.func.isRequired,
};

TimeForm.defaultProps = {
    formData: {time_settings: {}},
    formErrors: {time_settings: {}},
    setFormValue: () => {
    },
};

export default function TimeForm({ws, formData, formErrors, setFormValue, ...props}) {
    const [ntpData, triggerNTP] = useNTPDate(ws);
    useEffect(() => {
        if (ntpData.data) {
            const time = ntpData.data.time;
            const momentTime = moment(time).isValid() ? moment(time).format(TIME_FORMAT) : time;
            setFormValue(
                value => ({time_settings: {time: {$set: value}}})
            )({target: {value: momentTime}})
        }
    }, [setFormValue, ntpData.data]);

    function updateTimeFromBrowser(e) {
        e.preventDefault();
        setFormValue(value => (
            {time_settings: {time: {$set: value}}})
        )({target: {value: moment()}})
    }

    const data = formData.time_settings;
    const errors = formErrors.time_settings || {};

    function onDataTimeChangeHandler(value) {
        // Dirty hack to get DataTime library work
        if (typeof value === 'string')
            return setFormValue(
                value => ({time_settings: {time: {$set: value}}})
            )({target: {value: value}});
        return setFormValue(
            value => ({time_settings: {time: {$set: value}}})
        )({target: {value: value.format(TIME_FORMAT)}})
    }

    return <>
        <h4>{_('Time settings')}</h4>
        <p>{_('Time should be up-to-date otherwise DNS and other services might not work properly.')}</p>
        <Select
            label={_('How to set time')}
            choices={TIME_SETTING_TYPE_CHOICES}
            value={data.how_to_set_time}

            onChange={setFormValue(
                value => ({time_settings: {how_to_set_time: {$set: value}}})
            )}

            {...props}
        />
        {data.how_to_set_time === 'ntp' ? <NTPServersList servers={data.ntp_servers}/> : null}
        <DataTimeInput
            label={_('Time')}
            value={moment(data.time).isValid() ? moment(data.time) : data.time}
            error={errors.time}

            onChange={onDataTimeChangeHandler}

            {...props}

            disabled={data.how_to_set_time !== 'manual' || ntpData.isLoading}
        >
            <div className="input-group-append">
                <button
                    className="input-group-text"
                    onClick={data.how_to_set_time === 'ntp' ? triggerNTP : updateTimeFromBrowser}
                    disabled={ntpData.isLoading}
                >
                    {ntpData.isLoading ? <SpinnerElement small/> : <i className="fa fa-sync-alt"/>}
                </button>
            </div>
        </DataTimeInput>
    </>
}

NTPServersList.propTypes = {
    servers: PropTypes.arrayOf(PropTypes.string).isRequired
};

function NTPServersList({servers}) {
    const [shown, setShown] = useState(false);
    return <>
        <Button
            className="btn-outline-primary"
            forisFormSize
            data-toggle="collapse"
            href="#collapseNTPServers"
            onClick={e => {
                e.preventDefault();
                setShown(!shown)
            }}
        >
            {shown ? _('Hide NTP servers list') : _('Show NTP servers list')}
        </Button>

        <div className="collapse" id="collapseNTPServers">
            <h5>{_('NTP servers')}</h5>
            <div id="ntpServersList">
                {servers.map(server => <p key={server}>{server}</p>)}
            </div>
        </div>
    </>
}
