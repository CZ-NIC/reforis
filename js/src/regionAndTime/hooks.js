/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import {useEffect, useState} from 'react';
import {useAPIPost} from '../common/APIhooks';
import API_URLs from '../common/API';

export default function useNTPDate(ws) {
    const [state, setState] = useState({
        data: null,
        id: null,
        isLoading: false
    });

    const [updateTimeState, updateTime] = useAPIPost(API_URLs.ntpUpdate);
    useEffect(() => {
        if (updateTimeState.data) {
            setState((prevState,) => ({...prevState, id: updateTimeState.data.id}))
        }
    }, [updateTimeState.data]);

    function trigger() {
        setState(state => {
            return {...state, isLoading: true}
        });
        updateTime();
    }

    useEffect(() => {
        const module = 'time';
        ws.subscribe(module)
            .bind(module, 'ntpdate_started', msg => {
                if (state.id === msg.data.id)
                    setState((prevState,) => ({...prevState, isLoading: true}));
            })
            .bind(module, 'ntpdate_finished', msg => {
                if (state.id === msg.data.id)
                    setState((prevState,) => ({
                        ...prevState,
                        isLoading: false,
                        data: {time: msg.data.time, result: msg.data.result}
                    }));
            });
    }, [state.id, ws]);

    return [state, trigger]
}
