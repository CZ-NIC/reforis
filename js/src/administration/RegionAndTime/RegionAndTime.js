/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';
import moment from 'moment';


import RegionForm from './RegionForm';
import TimeForm from './TimeForm';
import ForisForm from '../../formContainer/ForisForm';

import {APIEndpoints} from '../../common/API';

export default function RegionAndTime() {
    return <>
        <h3>{_('Region and time')}</h3>
        <p>{_('It is important for your device to have the correct time set. If your device\'s time is delayed, the ' +
            'procedure of SSL certificate verification might not work correctly.')}</p>
        <ForisForm
            forisConfig={{
                endpoint: APIEndpoints.regionAndTime
            }}
            prepDataToSubmit={prepDataToSubmit}
            validator={validator}
        >
            <RegionForm/>
            <TimeForm/>
        </ForisForm>
    </>
}

function validator(formData) {
    if (!moment(formData.time_settings.time).isValid())
        return {time_settings: {time: _('Time should be in YYYY-MM-DD HH:MM:SS format.')}};
    return undefined;
}

function prepDataToSubmit(formData) {
    delete formData.time_settings.ntp_servers;
    if (formData.time_settings.how_to_set_time === 'ntp')
        delete formData.time_settings.time;
    return formData
}
