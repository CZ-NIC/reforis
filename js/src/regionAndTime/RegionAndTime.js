/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";
import moment from "moment";

import API_URLs from "common/API";
import { ForisForm } from "foris";

import RegionForm from "./RegionForm";
import TimeForm from "./TimeForm";

RegionAndTime.propTypes = {
    ws: PropTypes.object.isRequired,
    postCallback: PropTypes.func,
};

RegionAndTime.defaultProps = {
    postCallback: () => undefined,
};

export default function RegionAndTime({ ws, postCallback }) {
    return (
        <>
            <h1>{_("Region and Time")}</h1>
            <p>
                {_(`
It is important for your device to have the correct time set. If your device's time is delayed, the
procedure of SSL certificate verification might not work correctly.
        `)}
            </p>
            <ForisForm
                forisConfig={{
                    endpoint: API_URLs.regionAndTime,
                }}
                prepDataToSubmit={prepDataToSubmit}
                postCallback={postCallback}
                validator={validator}
            >
                <RegionForm />
                <TimeForm ws={ws} />
            </ForisForm>
        </>
    );
}

function validator(formData) {
    if (!moment(formData.time_settings.time).isValid()) return { time_settings: { time: _("Time should be in YYYY-MM-DD HH:MM:SS format.") } };
    return undefined;
}

function prepDataToSubmit(formData) {
    delete formData.time_settings.ntp_servers;
    if (formData.time_settings.how_to_set_time === "ntp") delete formData.time_settings.time;
    return formData;
}
