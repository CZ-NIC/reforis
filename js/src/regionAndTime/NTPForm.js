/*
 * Copyright (C) 2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { Button, TextInput } from "foris";
import { useNTPForm } from "./hooks";

export default function NTPForm({ formData, setShown }) {
    const [formState, setFormValue, saveServer] = useNTPForm(
        formData,
        setShown
    );

    return (
        <>
            <TextInput
                label={_("NTP Server")}
                onChange={setFormValue((value) => ({
                    newServer: { $set: value },
                }))}
            />
            <Button
                onClick={saveServer}
                forisFormSize
                className="btn-primary col-lg-12 col-md-12"
            >
                {_("Save")}
            </Button>
        </>
    );
}
