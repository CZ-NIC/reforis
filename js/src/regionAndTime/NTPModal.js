/*
 * Copyright (C) 2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import { Modal, ModalBody, ModalHeader } from "foris";
import NTPForm from "./NTPForm";

export default function NTPModal({
    shown,
    setShown,
    title,
    servers,
    formData,
}) {
    return (
        <Modal scrollable shown={shown} setShown={setShown}>
            <ModalHeader setShown={setShown} title={title} />
            <ModalBody>
                <NTPForm servers={servers} formData={formData} />
            </ModalBody>
        </Modal>
    );
}
