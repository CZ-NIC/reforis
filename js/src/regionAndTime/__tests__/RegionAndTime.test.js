/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from 'react';

import {getByText, render, wait} from 'customTestRender';
import mockAxios from 'jest-mock-axios';
import {regionAndTime} from './__fixtures__/regionAndTime';

import RegionAndTime from '../RegionAndTime';
import API_URLs from '../../common/API';

describe('<RegionAndTime/>', () => {
    let regionAndTimeContainer;

    beforeEach(async () => {
        const {container} = render(<RegionAndTime/>);
        mockAxios.mockResponse({data: regionAndTime()});
        await wait(() => getByText(container, 'Region settings'));
        regionAndTimeContainer = container;
    });

    it('Snapshot', () => {
        expect(mockAxios.get).toHaveBeenCalledTimes(1);
        expect(mockAxios.get).toHaveBeenCalledWith(API_URLs.regionAndTime, expect.anything(),);
        expect(regionAndTimeContainer).toMatchSnapshot();
    });
});
