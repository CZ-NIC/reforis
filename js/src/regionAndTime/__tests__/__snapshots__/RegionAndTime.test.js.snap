// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegionAndTime/> Snapshot 1`] = `
<div>
  <h1>
    Region and time
  </h1>
  <p>
    
It is important for your device to have the correct time set. If your device's time is delayed, the
procedure of SSL certificate verification might not work correctly.
        
  </p>
  <form>
    <h4>
      Region settings
    </h4>
    <p>
      Please select the timezone the router is being operated in. Correct setting is required to display the right time and for related functions.
    </p>
    <div
      class="form-group col-sm-12 offset-lg-1 col-lg-10"
    >
      <label
        for="1"
      >
        Continent or ocean
      </label>
      <select
        class="custom-select"
        id="1"
      >
        <option
          value="Africa"
        >
          Africa
        </option>
        <option
          value="America"
        >
          America
        </option>
        <option
          value="Antarctica"
        >
          Antarctica
        </option>
        <option
          value="Arctic"
        >
          Arctic
        </option>
        <option
          value="Asia"
        >
          Asia
        </option>
        <option
          value="Atlantic"
        >
          Atlantic
        </option>
        <option
          value="Australia"
        >
          Australia
        </option>
        <option
          value="Europe"
        >
          Europe
        </option>
        <option
          value="Indian"
        >
          Indian
        </option>
        <option
          value="Pacific"
        >
          Pacific
        </option>
      </select>
    </div>
    <div
      class="form-group col-sm-12 offset-lg-1 col-lg-10"
    >
      <label
        for="2"
      >
        Country
      </label>
      <select
        class="custom-select"
        id="2"
      >
        <option
          value="AD"
        >
          Andorra
        </option>
        <option
          value="AL"
        >
          Albania
        </option>
        <option
          value="AT"
        >
          Austria
        </option>
        <option
          value="AX"
        >
          Åland Islands
        </option>
        <option
          value="BA"
        >
          Bosnia & Herzegovina
        </option>
        <option
          value="BE"
        >
          Belgium
        </option>
        <option
          value="BG"
        >
          Bulgaria
        </option>
        <option
          value="BY"
        >
          Belarus
        </option>
        <option
          value="CH"
        >
          Switzerland
        </option>
        <option
          value="CZ"
        >
          Czech Republic
        </option>
        <option
          value="DE"
        >
          Germany
        </option>
        <option
          value="DK"
        >
          Denmark
        </option>
        <option
          value="EE"
        >
          Estonia
        </option>
        <option
          value="ES"
        >
          Spain
        </option>
        <option
          value="FI"
        >
          Finland
        </option>
        <option
          value="FR"
        >
          France
        </option>
        <option
          value="GB"
        >
          United Kingdom
        </option>
        <option
          value="GG"
        >
          Guernsey
        </option>
        <option
          value="GI"
        >
          Gibraltar
        </option>
        <option
          value="GR"
        >
          Greece
        </option>
        <option
          value="HR"
        >
          Croatia
        </option>
        <option
          value="HU"
        >
          Hungary
        </option>
        <option
          value="IE"
        >
          Ireland
        </option>
        <option
          value="IM"
        >
          Isle of Man
        </option>
        <option
          value="IT"
        >
          Italy
        </option>
        <option
          value="JE"
        >
          Jersey
        </option>
        <option
          value="LI"
        >
          Liechtenstein
        </option>
        <option
          value="LT"
        >
          Lithuania
        </option>
        <option
          value="LU"
        >
          Luxembourg
        </option>
        <option
          value="LV"
        >
          Latvia
        </option>
        <option
          value="MC"
        >
          Monaco
        </option>
        <option
          value="MD"
        >
          Moldova
        </option>
        <option
          value="ME"
        >
          Montenegro
        </option>
        <option
          value="MK"
        >
          Macedonia
        </option>
        <option
          value="MT"
        >
          Malta
        </option>
        <option
          value="NL"
        >
          Netherlands
        </option>
        <option
          value="NO"
        >
          Norway
        </option>
        <option
          value="PL"
        >
          Poland
        </option>
        <option
          value="PT"
        >
          Portugal
        </option>
        <option
          value="RO"
        >
          Romania
        </option>
        <option
          value="RS"
        >
          Serbia
        </option>
        <option
          value="RU"
        >
          Russia
        </option>
        <option
          value="SE"
        >
          Sweden
        </option>
        <option
          value="SI"
        >
          Slovenia
        </option>
        <option
          value="SK"
        >
          Slovakia
        </option>
        <option
          value="SM"
        >
          San Marino
        </option>
        <option
          value="TR"
        >
          Turkey
        </option>
        <option
          value="UA"
        >
          Ukraine
        </option>
        <option
          value="VA"
        >
          Vatican City
        </option>
      </select>
    </div>
    <div
      class="form-group col-sm-12 offset-lg-1 col-lg-10"
    >
      <label
        for="3"
      >
        Timezone
      </label>
      <select
        class="custom-select"
        id="3"
      >
        <option
          value="Prague"
        >
          Prague
        </option>
      </select>
    </div>
    <h4>
      Time settings
    </h4>
    <p>
      Time should be up-to-date otherwise DNS and other services might not work properly.
    </p>
    <div
      class="form-group col-sm-12 offset-lg-1 col-lg-10"
    >
      <label
        for="4"
      >
        How to set time
      </label>
      <select
        class="custom-select"
        id="4"
      >
        <option
          value="ntp"
        >
          Via NTP
        </option>
        <option
          value="manual"
        >
          Manually
        </option>
      </select>
    </div>
    <button
      class="btn btn-outline-primary offset-lg-8 col-lg-3 col-sm-12"
      data-toggle="collapse"
      href="#collapseNTPServers"
      type="button"
    >
       
       
      Show NTP servers list
    </button>
    <div
      class="collapse"
      id="collapseNTPServers"
    >
      <h5>
        NTP servers
      </h5>
      <div
        id="ntpServersList"
      >
        <p>
          217.31.202.100
        </p>
        <p>
          195.113.144.201
        </p>
        <p>
          195.113.144.238
        </p>
        <p>
          2001:1488:ffff::100
        </p>
        <p>
          ntp.nic.cz
        </p>
        <p>
          0.openwrt.pool.ntp.org
        </p>
        <p>
          1.openwrt.pool.ntp.org
        </p>
        <p>
          2.openwrt.pool.ntp.org
        </p>
        <p>
          3.openwrt.pool.ntp.org
        </p>
      </div>
    </div>
    <div
      class="rdt"
    >
      <div>
        <div
          class="col-sm-12 offset-lg-1 col-lg-10"
        >
          <div
            class="form-group"
          >
            <label
              for="5"
            >
              Time
            </label>
            <div
              class="input-group"
            >
              <input
                class="form-control form-control"
                disabled=""
                id="5"
                type="text"
                value="2019-04-24 21:15:03"
              />
              <div
                class="input-group-append"
              >
                <button
                  class="input-group-text"
                >
                  <i
                    class="fa fa-sync-alt"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="rdtPicker"
      >
        <div
          class="rdtDays"
        >
          <table>
            <thead>
              <tr>
                <th
                  class="rdtPrev"
                >
                  <span>
                    ‹
                  </span>
                </th>
                <th
                  class="rdtSwitch"
                  colspan="5"
                  data-value="3"
                >
                  April 2019
                </th>
                <th
                  class="rdtNext"
                >
                  <span>
                    ›
                  </span>
                </th>
              </tr>
              <tr>
                <th
                  class="dow"
                >
                  Su
                </th>
                <th
                  class="dow"
                >
                  Mo
                </th>
                <th
                  class="dow"
                >
                  Tu
                </th>
                <th
                  class="dow"
                >
                  We
                </th>
                <th
                  class="dow"
                >
                  Th
                </th>
                <th
                  class="dow"
                >
                  Fr
                </th>
                <th
                  class="dow"
                >
                  Sa
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td
                  class="rdtDay rdtOld"
                  data-value="31"
                >
                  31
                </td>
                <td
                  class="rdtDay"
                  data-value="1"
                >
                  1
                </td>
                <td
                  class="rdtDay"
                  data-value="2"
                >
                  2
                </td>
                <td
                  class="rdtDay"
                  data-value="3"
                >
                  3
                </td>
                <td
                  class="rdtDay"
                  data-value="4"
                >
                  4
                </td>
                <td
                  class="rdtDay"
                  data-value="5"
                >
                  5
                </td>
                <td
                  class="rdtDay"
                  data-value="6"
                >
                  6
                </td>
              </tr>
              <tr>
                <td
                  class="rdtDay"
                  data-value="7"
                >
                  7
                </td>
                <td
                  class="rdtDay"
                  data-value="8"
                >
                  8
                </td>
                <td
                  class="rdtDay"
                  data-value="9"
                >
                  9
                </td>
                <td
                  class="rdtDay"
                  data-value="10"
                >
                  10
                </td>
                <td
                  class="rdtDay"
                  data-value="11"
                >
                  11
                </td>
                <td
                  class="rdtDay"
                  data-value="12"
                >
                  12
                </td>
                <td
                  class="rdtDay"
                  data-value="13"
                >
                  13
                </td>
              </tr>
              <tr>
                <td
                  class="rdtDay"
                  data-value="14"
                >
                  14
                </td>
                <td
                  class="rdtDay"
                  data-value="15"
                >
                  15
                </td>
                <td
                  class="rdtDay"
                  data-value="16"
                >
                  16
                </td>
                <td
                  class="rdtDay"
                  data-value="17"
                >
                  17
                </td>
                <td
                  class="rdtDay"
                  data-value="18"
                >
                  18
                </td>
                <td
                  class="rdtDay"
                  data-value="19"
                >
                  19
                </td>
                <td
                  class="rdtDay"
                  data-value="20"
                >
                  20
                </td>
              </tr>
              <tr>
                <td
                  class="rdtDay"
                  data-value="21"
                >
                  21
                </td>
                <td
                  class="rdtDay"
                  data-value="22"
                >
                  22
                </td>
                <td
                  class="rdtDay"
                  data-value="23"
                >
                  23
                </td>
                <td
                  class="rdtDay rdtActive"
                  data-value="24"
                >
                  24
                </td>
                <td
                  class="rdtDay"
                  data-value="25"
                >
                  25
                </td>
                <td
                  class="rdtDay"
                  data-value="26"
                >
                  26
                </td>
                <td
                  class="rdtDay"
                  data-value="27"
                >
                  27
                </td>
              </tr>
              <tr>
                <td
                  class="rdtDay"
                  data-value="28"
                >
                  28
                </td>
                <td
                  class="rdtDay"
                  data-value="29"
                >
                  29
                </td>
                <td
                  class="rdtDay"
                  data-value="30"
                >
                  30
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="1"
                >
                  1
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="2"
                >
                  2
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="3"
                >
                  3
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="4"
                >
                  4
                </td>
              </tr>
              <tr>
                <td
                  class="rdtDay rdtNew"
                  data-value="5"
                >
                  5
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="6"
                >
                  6
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="7"
                >
                  7
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="8"
                >
                  8
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="9"
                >
                  9
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="10"
                >
                  10
                </td>
                <td
                  class="rdtDay rdtNew"
                  data-value="11"
                >
                  11
                </td>
              </tr>
            </tbody>
            <tfoot>
              <tr>
                <td
                  class="rdtTimeToggle"
                  colspan="7"
                >
                  21:15:03
                </td>
              </tr>
            </tfoot>
          </table>
        </div>
      </div>
    </div>
    <button
      class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
      type="submit"
    >
       
       
      Save
    </button>
  </form>
</div>
`;
